plugins/woocommerce/src/StoreApi/Routes/V1/Checkout.php

```diff
@@ -177,58 +177,138 @@ protected function get_route_response( \WP_REST_Request $request ) {
 	}
 
 	/**
-	 * Validate required additional fields on request.
+	 * Validation callback for the checkout route.
 	 *
-	 * @param \WP_REST_Request $request Request object.
+	 * This runs after individual field validation_callbacks have been called.
 	 *
-	 * @throws RouteException When a required additional field is missing.
+	 * @param \WP_REST_Request $request Request object.
+	 * @return true|\WP_Error
 	 */
-	public function validate_required_additional_fields( \WP_REST_Request $request ) {
-		$document_object = null;
+	public function validate_callback( $request ) {
+		$sanitized_request = clone $request;
+		$sanitized_request->sanitize_params();
 
 		if ( Features::is_enabled( 'experimental-blocks' ) ) {
 			$document_object = new DocumentObject(
 				[
 					'customer' => [
-						'billing_address'  => $request['billing_address'],
-						'shipping_address' => $request['shipping_address'],
+						'billing_address'   => $sanitized_request->get_param( 'billing_address' ),
+						'shipping_address'  => $sanitized_request->get_param( 'shipping_address' ),
+						'additional_fields' => $this->additional_fields_controller->filter_values_for_location(
+							$sanitized_request->get_param( 'additional_fields' ) ?? [],
+							'contact'
+						),
 					],
 					'checkout' => [
-						'payment_method'    => $request['payment_method'],
-						'create_account'    => $request['create_account'],
-						'customer_note'     => $request['customer_note'],
-						'additional_fields' => $request['additional_fields'],
+						'payment_method'    => $sanitized_request->get_param( 'payment_method' ),
+						'create_account'    => $sanitized_request->get_param( 'create_account' ),
+						'customer_note'     => $sanitized_request->get_param( 'customer_note' ),
+						'additional_fields' => $this->additional_fields_controller->filter_values_for_location(
+							$sanitized_request->get_param( 'additional_fields' ) ?? [],
+							'order'
+						),
 					],
 				]
 			);
+		} else {
+			$document_object = null;
 		}
 
-		$address_fields = $this->additional_fields_controller->get_fields_for_location( 'address' );
+		$validate_contexts = [
+			'shipping_address' => [
+				'group'    => 'shipping',
+				'location' => 'address',
+				'param'    => 'shipping_address',
+			],
+			'billing_address'  => [
+				'group'    => 'billing',
+				'location' => 'address',
+				'param'    => 'billing_address',
+			],
+			'contact'          => [
+				'group'    => 'other',
+				'location' => 'contact',
+				'param'    => 'additional_fields',
+			],
+			'order'            => [
+				'group'    => 'other',
+				'location' => 'order',
+				'param'    => 'additional_fields',
+			],
+		];
```

Comment:
Do you think those should live in a trait or somewhere else? they're majorly distracting the flow.